p2p/chain.py

```diff
@@ -164,142 +149,134 @@ def register_peer(self, peer: BasePeer) -> None:
+        batches = list(partition_all(length, headers))
+        for peer, batch in zip(self.peer_pool.peers, batches):
+            request_func(cast(ETHPeer, peer), batch)
+        return len(batches)
+
+    def _send_get_block_bodies(self, peer: ETHPeer, headers: List[BlockHeader]) -> None:
         self.logger.debug("Requesting %d block bodies to %s", len(headers), peer)
-        now = time.time()
-        for header in headers:
-            key = (header.transaction_root, header.uncles_hash)
-            self._pending_bodies[key] = (header, now)
-
-    @to_list
-    def _skip_empty_and_duplicated_receipts(
-            self, headers: List[BlockHeader]) -> Generator[BlockHeader, None, None]:
-        # Post-Byzantium blocks may have identical receipt roots (e.g. when they have the same
-        # number of transactions and all succeed/failed: ropsten blocks 2503212 and 2503284), so
-        # we have an extra check here to avoid requesting those receipts multiple times.
-        headers = list(unique(headers, key=operator.attrgetter('receipt_root')))
-        for header in headers:
-            if (header.receipt_root != self.chaindb.empty_root_hash and
-                    header.receipt_root not in self._pending_receipts):
-                yield header
-
-    async def request_receipts(self, headers: List[BlockHeader]) -> None:
-        peer = await self.get_idle_peer()
-        peer.sub_proto.send_get_receipts([header.hash for header in headers])
-        self._peers_with_pending_requests.add(peer)
+        peer.sub_proto.send_get_block_bodies([header.hash for header in headers])
+
+    def _send_get_receipts(self, peer: ETHPeer, headers: List[BlockHeader]) -> None:
         self.logger.debug("Requesting %d block receipts to %s", len(headers), peer)
-        now = time.time()
-        for header in headers:
-            self._pending_receipts[header.receipt_root] = (header, now)
+        peer.sub_proto.send_get_receipts([header.hash for header in headers])
 
-    async def get_idle_peer(self) -> ETHPeer:
-        """Return a random peer which we're not already expecting a response from."""
-        while True:
-            idle_peers = [
-                peer
-                for peer in self.peer_pool.peers
-                if peer not in self._peers_with_pending_requests
-            ]
-            if idle_peers:
-                return cast(ETHPeer, secrets.choice(idle_peers))
-            else:
-                self.logger.debug("No idle peers availabe, sleeping a bit")
-                await asyncio.sleep(0.2)
+    def _body_key(self, header: BlockHeader) -> Tuple[bytes, bytes]:
+        """Return the unique key of the body for the given header.
+
+        i.e. a two-tuple with the transaction root and uncles hash.
+        """
+        return cast(Tuple[bytes, bytes], (header.transaction_root, header.uncles_hash))
+
+    def _is_body_empty(self, header: BlockHeader) -> bool:
```

Comment:
Same (elevate to top level module function)

Reply:
In order to more this one I'd need to add an extra argument (for the empty_root_hash, which it currently takes from the chaindb), so I'd rather keep it here unless you feel strongly about it

Reply:
You can actually hard code that value as the alternate trie implementation which required the empty root hash to be specified is being removed as part of #518